### ISSUE TITLE
issue assigning aggregates in outliner

### ISSUE THREAD

**Issue by Andrej730** [IfcOpenShell/IfcOpenShell#4763]
There is an issue that occurs sometimes when you try to reassign one aggregate to another my moving objects in outliner and save the project.

https://github.com/IfcOpenShell/IfcOpenShell/assets/9417531/80c48db9-6104-45ef-a00d-74b1959818ac

Basically the problem is that of syncing objects at saving the problem depends on the order they were internal mapping we have:
https://github.com/IfcOpenShell/IfcOpenShell/blob/3679fccfbd5f5d94cddfdacdda5bb96e402c7781/src/blenderbim/blenderbim/bim/export_ifc.py#L95

In the example above I'm trying to move a building from site "My Site" to "Site 2" but since so happened that IfcBuildingStorey comes in the mapping first, `tool.Collector.sync` is running `blenderbim.core.aggregate.assign_object` for it which moved relating object (IfcBuilding) back to the "My Site".

https://github.com/IfcOpenShell/IfcOpenShell/blob/45a16554857336d11b512eed30dc78b1719ccf9a/src/blenderbim/blenderbim/core/aggregate.py#L36-L50

Do we really need to run `collector.assign(relating_obj)` here? E.g. when we do `spatial.assign_container` we assign collection only for the related object.
https://github.com/IfcOpenShell/IfcOpenShell/blob/45a16554857336d11b512eed30dc78b1719ccf9a/src/blenderbim/blenderbim/core/spatial.py#L61-L64

### how to reproduce the issue

It does occur randomly from time to time with any aggregates and I've found some (rather hacky) way to reproduce it by the file and script below (run the script, move ifcbuilding and save the project to reproduce the issue).

[example.ifc.txt](https://github.com/IfcOpenShell/IfcOpenShell/files/15484618/example.ifc.txt)

```python
from blenderbim.bim.ifc import IfcStore
import bpy
id_map = {
    42: bpy.data.objects["IfcBuildingStorey/My Storey"],
    99: bpy.data.objects["IfcBuilding/Building 2"],
    1: bpy.data.objects["IfcProject/My Project"],
    30: bpy.data.objects["IfcSite/My Site"],
    66: bpy.data.objects["IfcSite/Site 2"],
}
IfcStore.id_map = id_map
```

**Comment by Andrej730** [comment]
Not sure about other UI options, will need to investigate, would be nice to avoid rebuilding the outliner if possible 😁

If only there was a some way in Blender Python API to hook up to the event after moving items in outliner, then we wouldn't need to sync anything or atleast it will be possible not to sync everything each time.

But just to add to the problems with syncing - when you save large projects syncing is also taking a lot of time even if you haven't changed anything.

**Comment by Moult** [comment]
Yeah syncing is slow.

I wouldn't rebuild the outliner, I'd say most users want to see either the spatial tree, custom groups, types, or some other special grouping. For larger projects, seeing the object tree of every single object is really unwieldy, the first thing I always do is shift click to collapse it.

It's also really weird how aggregation, containment and the spatial manager are in one part of the UI, but the outliner is in another.

Dion Moult

Sent from Proton Mail mobile

-------- Original Message --------

**Comment by Moult** [comment]
There's all sorts of broken behavior you can trigger from the outliner. Like moving stuff into drawing groups, creating new groups, or moving spatial elements into physical elements, or having multiple collection parents...

It just seems safer in the future to decouple from collections. Let users do whatever they want with collections.

**Comment by Andrej730** [comment]
I think the main benefits from outliner are 

1) drag'n'drop is very convenient (and there seems to be no other ui elements with drag'n-drop - https://blender.stackexchange.com/questions/291672/drag-and-drop-item-in-template-list); 

2) everyone is more familiar with concept of outliner. Even for users who are not familiar with Blender but familar with other software, intuitively it seem to make sense that it should be possible to change hierarchy using drag'n'drop in outliner. If we'll take it away from the outliner, then to avoid the confusion we'll need some way to communicate in UI that outliner hierarchy is completely unrelated to the current IFC hierarchy.

But yes, since there is no way to hook up to the changes in outliner and therefore we can't prevent any unsafe actions, it seems that we'll need to move away from outline eventually if there be will no workaround.

Ideally, what kind of event we would need from Blender to make it work - just `obj_or_col_collection_changed(obj_or_col, prev_collection, new_collection)`? In theory, even if there was `obj_or_col_collection_changed(obj_or_col)` we can keep track of the changes and prevent unsafe ones.

**Comment by Moult** [comment]
Collections also slow down the load process. Especially for steel models which have tons of assemblies, basically take every beam and multiply by 2.

I vaguely recall work by @ecke101 / erik85 upstream to get collection change callbacks but it was unsuccessful iirc and I can't find it now. I'm not convinced it'll solve all the problems either.

There's some promising work upstream on treeviews... https://developer.blender.org/docs/features/interface/views/tree_views/ and https://projects.blender.org/blender/blender/pulls/109825

Drag and drop is nice but I think the outliner is overrated once the project grows larger. Seeing 200 IfcWalls all named Wall doesn't help. Smarter aggregation by type can be done with a custom UI.

I'm strongly considering replacing the outliner with another properties panel, and promoting the spatial manager to a more powerful tool.

**Comment by Andrej730** [comment]
Another downside of the outliner - since we planned not to force on users collections arrangements in outliner so they could setup whatever they find convenient, then it will be more and more trickier to keep it safe even if there was a way to be notified about outliner changes.

> There's some promising work upstream on treeviews... [developer.blender.org/docs/features/interface/views/tree_views](https://developer.blender.org/docs/features/interface/views/tree_views/) and [projects.blender.org/blender/blender/pulls/109825](https://projects.blender.org/blender/blender/pulls/109825)

That looks like a perfect solution if it will be available from Python API at some point.

**Comment by Andrej730** [comment]
Resolved after e3cb811
